### PR TITLE
Remove failures on ebpf_hash_table_delete path due to low memory

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Resolves: #1506 

## Description

The ebpf_hash_table_delete API can fail in out-of-memory scenarios. To resolve this, pre-allocate the replacement block required for removing an element when adding an element to the hash table.

Each hash table bucket entry located at index N contains a pointer to memory needed to create a replacement bucket containing N-1 entries. 

## Testing

CI/CD

## Documentation

No.

### CODE REVIEW

**Comment by Alan-Jowett** [comment]
This change doesn't appear to have a statistically significant impact on the performance of hash table operations.

Baseline:
https://github.com/microsoft/ebpf-for-windows/actions/runs/3364756454/jobs/5579741677#step:14:28

This PR:
https://github.com/microsoft/ebpf-for-windows/actions/runs/3370589012/jobs/5591999415#step:14:28

File: libs/platform/ebpf_hash_table.c [hunk 1]
```diff
@@ -214,9 +239,195 @@ _ebpf_hash_table_bucket_entry(size_t key_size, _In_ ebpf_hash_bucket_header_t* b
...
+    memcpy(entry->key, key, hash_table->key_size);
+    local_new_bucket->count++;
+
+    *new_bucket = local_new_bucket;
+    local_new_bucket = NULL;
+
+    result = EBPF_SUCCESS;
+
+Done:
+    ebpf_free(local_new_bucket);
```
Comment (Alan-Jowett) [inline libs/platform/ebpf_hash_table.c:307]: Wrong free. Should be hash_table->free

**Comment by dthaler** [comment]
This PR isn't currently linked to issue 1506, should it be?

File: libs/platform/ebpf_hash_table.c [hunk 2]
```diff
@@ -214,9 +239,194 @@ _ebpf_hash_table_bucket_entry(size_t key_size, _In_ ebpf_hash_bucket_header_t* b
...
+        if (index == key_index) {
+            continue;
+        }
+        const ebpf_hash_bucket_entry_t* old_entry =
+            _ebpf_hash_table_bucket_entry(hash_table->key_size, old_bucket, index);
+        ebpf_hash_bucket_entry_t* new_entry =
+            _ebpf_hash_table_bucket_entry(hash_table->key_size, backup_bucket, backup_bucket->count);
+
+        new_entry->data = old_entry->data;
+        memcpy(new_entry->key, old_entry->key, hash_table->key_size);
```
Comment (saxena-anurag) [inline libs/platform/ebpf_hash_table.c:357]: If say the bucket had 10 entries (index 0-9), each entry will have a backup bucket of size n-1 where n is the index of the entry.

If we delete, for example, entry at index 5, then entries at index 6-9 will be shifted left in the new bucket, and their indices will decrease by 1. But their backup buckets still be of a larger size. 

Will that cause any issues?

File: libs/platform/ebpf_hash_table.c [hunk 3]
```diff
@@ -13,37 +13,62 @@
 // searching, data is stored separately to prevent read-copy-update semantics
 // from causing loss of updates.
 
+/**
+ * @brief Each bucket entry contains a pointer to the value, the key and a pointer to pre-allocated memory that can be
+ * used to replace the current bucket with a bucket one entry smaller.
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply libs/platform/ebpf_hash_table.c:18]: Assume there are 3 entries in the bucket.
Entry 3 has a backup_bucket of size 2 stored in it.
Entry 2 has a backup_bucket of size 1 stored in it.
Entry 1 has no backup_bucket (a zero-sized backup_bucket would be redundant).

When the first delete occurs (doesn't matter where from), the backup_bucket from entry 3 is used to build the new bucket.
When the second delete occurs the backup_bucket from entry 2 is used to build the new bucket.
When the third and final delete occurs, the hash bucket is set to NULL.

File: libs/platform/ebpf_hash_table.c (hunk 2 above)
Comment (Alan-Jowett) [reply libs/platform/ebpf_hash_table.c:357]: See the lines below this. Backup_buckets for entries 1 through N-1 are copied from old_bucket to new_bucket, in order.

The assert below in the code:
```
        ebpf_assert(index == 0 || new_entry->backup_bucket);
        ebpf_assert(index == 0 || new_entry->backup_bucket->count == index);
```
Verifies that each entry has an appropriately sized bucket associated with it.

File: libs/platform/ebpf_hash_table.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply libs/platform/ebpf_hash_table.c:18]: Note:
This means that a bucket containing N entries has N-1 backup buckets and can handle N-1 deletions without additional allocations. If an entry is inserted, it allocates both the new bucket and a new backup bucket.

**Review by saxena-anurag — APPROVED** [review]

**Review by dthaler — APPROVED** [review]
Please link PR to issue 1506 if this fixes it.

File: libs/platform/ebpf_hash_table.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply libs/platform/ebpf_hash_table.c:18]: https://github.com/microsoft/ebpf-for-windows/issues/1552

**Pull request merged**